azure-iot-device/azure/iot/device/iothub/aio/loop_management.py

```diff
@@ -0,0 +1,72 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+""" This module contains functions of managing event loops for the IoTHub client
+"""
+import asyncio
+import threading
+import logging
+from azure.iot.device.common import asyncio_compat
+
+logger = logging.getLogger(__name__)
+
+loops = {
+    # Store whatever loop the user has on their thread the client was created in.
+    # We can use this to schedule tasks for their handler/callback code so that any
+    # poor performance in their provided code doesn't slow down the client.
+    "USER_LOOP": asyncio.get_event_loop(),
+    "CLIENT_INTERNAL_LOOP": None,
+    "CLIENT_HANDLER_RUNNER_LOOP": None,
+}
+
+
+def _cleanup():
+    """Clear all running loops and end respective threads.
+    Does not clear the USER_LOOP.
+    ONLY FOR TESTING USAGE
+    By using this function, you can wipe all global loops.
+    DO NOT USE THIS IN PRODUCTION CODE
+    """
+    for loop_name, loop in loops.items():
+        if loop_name == "USER_LOOP":
+            # Do not clean up the USER_LOOP since it wasn't made by us
+            continue
+        elif loop is not None:
+            loop.call_soon_threadsafe(loop.stop())
+            # NOTE: Stopping the loop will also end the thread, because the only thing keeping
+            # the thread alive was the loop running
+            loops[loop_name] = None
+
+
+def _make_new_loop(loop_name):
+    logger.debug("Creating new event loop - {}".format(loop_name))
+    # Create the loop on a new Thread
+    new_loop = asyncio.new_event_loop()
+    loop_thread = threading.Thread(target=new_loop.run_forever)
```

Comment:
I think this is in the wrong order.  Instead of crating an event loop in the current thread and then running it in a second thread, I think you want to crate a second thread and then create and run the event loop inside the second thread.

I was just reading the implementations for 3 of the 4 "included" loop policies, and I wasn't able to find a any problems with doing it your way, but it just feels like you should create the loop from inside the thread that it's running in.

But -- and I don't think we're going to get this right for all possible cases -- this code is built on the notion of one loop per thread, but that is not necessarily always the case.  The documentation for event loop policies ([https://docs.python.org/3/library/asyncio-policy.html](https://docs.python.org/3/library/asyncio-policy.html)) says "A policy defines the notion of context and manages a separate event loop per context. The default policy defines context to be the current thread."  This means that 1:1 mapping between threads and event loops is the _default_ but it's not guaranteed.  The asyncio implementation talks about theoretical policies that might use different contexts:

```
class BaseDefaultEventLoopPolicy(AbstractEventLoopPolicy):
    """Default policy implementation for accessing the event loop.

    In this policy, each thread has its own event loop.  However, we
    only automatically create an event loop by default for the main
    thread; other threads by default have no event loop.

    Other policies may have different rules (e.g. a single global
    event loop, or automatically creating an event loop per thread, or
    using some other notion of context to which an event loop is
    associated).
    """
```

This gets into brain-twisting territory.  If there's only one event loop for all threads, then how does `new_event_loop` behave?  It might be safer to call `get_event_loop` instead because that one will either return the current one (if it exists for the current context) or create a new one (if there is no event loop for the current context), and this would work for the default case and the "one event loop shared between all threads" case.  But, if there's a singleton event loop that's shared between all threads, what happens if you call `run_forever` on it when it's already running?

I know that pytest-asyncio made some accommodations for custom event loop policies, but I don't know if they fiddle with the policy or if they're just careful about how they use asyncio because there might be a custom policy in play.  

I think I just bumped my head on one of the rocks down here in this asyncio rabbit hole.  damn.

Reply:
A few things:

1) I don't see a benefit to creating the loop in another thread, because that just makes things more complicated. After all, a Thread simply runs a single target function until execution is completed. If we had to build a new loop in a separate thread, and then run it, that's two operations, which would mean a custom function would need to be defined in order to do those two things (+ store it in the dictionary), and I just don't see the benefit to doing that. I think simply running it in a separate thread is simpler.

2) Given the low-level manual loop creation we're doing here, I don't think loop policy should make too much of an impact. Calling `get_event_loop` will return the default loop based on policy, but we're operating at a lower level here by making our own. And since `new_event_loop()` is part of the abstract policy, it will always be implemented no matter what custom policy is in effect. It also cannot return None. I also don't know what this means for the case of a policy enforcing a singular global loop. My feeling based on reading the docs is that as long as we're manually dealing with loops ourselves, we should probably be fine, a lot of the policy stuff only comes into effect when we try to use default loops for a context, which isn't something that happens here.

---
In reply to: [480468732](https://github.com/Azure/azure-iot-sdk-python/pull/651#discussion_r480468732) [](ancestors = 480468732)